src/Microsoft.Extensions.Logging.MSBuild/MSBuildLogger.cs

```diff
@@ -0,0 +1,257 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Microsoft.NET.StringTools;
+
+namespace Microsoft.Extensions.Logging.MSBuild;
+
+/// <summary>
+/// Implements an ILogger that passes the logs to the wrapped TaskLoggingHelper.
+/// </summary>
+/// <remarks>
+/// This logger is designed to be used with MSBuild tasks, allowing logs to be written in a way that integrates with the MSBuild logging system.
+/// It looks for specific property names in the state/scope parts of the message and maps them to the parameters of the MSBuild LogX methods.
+/// Those specific keys are:
+/// <list type="bullet">
+/// <item><term>Subcategory</term></item>
+/// <item><term>Code</term></item>
+/// <item><term>HelpKeyword</term></item>
+/// <item><term>File</term></item>
+/// <item><term>LineNumber</term></item>
+/// <item><term>ColumnNumber</term></item>
+/// <item><term>EndLineNumber</term></item>
+/// <item><term>EndColumnNumber</term></item>
+/// <item><term>{OriginalFormat}</term><description>(usually provided by the underlying logging framework)</description></item>
+/// </list>
+///
+/// So if you add these to the scope (e.g. via <code lang="csharp">_logger.BeginScope(new Dictionary&lt;string, object&gt;{ ... }))</code> or on the message format itself,
+/// they will be extracted and used to format the message correctly for MSBuild.
+/// </remarks>
+public sealed class MSBuildLogger : ILogger
+{
+    private static readonly IDisposable Scope = new DummyDisposable();
+
+    private readonly TaskLoggingHelper _loggingHelper;
+    private readonly string _category;
+    private IExternalScopeProvider? _scopeProvider;
+
+    public MSBuildLogger(string category, TaskLoggingHelper loggingHelperToWrap, IExternalScopeProvider? scopeProvider = null)
+    {
+        _category = category;
+        _loggingHelper = loggingHelperToWrap;
+        _scopeProvider = scopeProvider;
+    }
+
+    IDisposable ILogger.BeginScope<TState>(TState state) => _scopeProvider?.Push(state) ?? Scope;
+
+    public bool IsEnabled(LogLevel logLevel) => true;
```

Comment:
This can be implemented to call `_loggingHelper.LogsMessagesOfImportance` if `logLevel` is less than warnning.

Reply:
good call! I've just done that